code/modules/mailorder/modr_catalogues.dm

```diff
@@ -0,0 +1,287 @@
+
+	chem
+		name = "Chems-r-Us"
+		entries_to_index = /datum/mail_order/chem
+
+	New()
+		..()
+		for(var/S in concrete_typesof(entries_to_index))
+			var/datum/mail_order/mo_entry = new S()
+			src.canbuy[mo_entry.cleanname] = mo_entry
+
+	return_text()
+		. = src.return_text_header()
+		. += " | <a href='byond://?src=\ref[src];viewlist=1'>Catalogue</a>"
+		if(length(src.cart) > 0)
+			var/cartlength = length(src.cart)
+			. += " | <a href='byond://?src=\ref[src];viewcart=1'>Cart ([cartlength])</a>"
+		else
+			. += " | <a href='byond://?src=\ref[src];viewcart=1'>Cart</a>"
+		. += "<h4>[src.name]</h4><hr>"
+
+		if(!src.master.host_program)
+			. += "ERROR 404: File not found."
+		switch(src.mode)
+			if(MODE_LIST)
+				. += "<h4>Catalogue</h4><br>"
+				if(length(src.canbuy) < 1)
+					. += "None!"
+				else
+					for(var/P in src.canbuy)
+						var/datum/mail_order/F = src.canbuy[P]
+						if(!istype(F))
+							continue
+						var/itemct = length(F.order_items)
+						. += {"<a href='byond://?src=\ref[src];add_to_cart=[F.cleanname]'>[F.name]</a> - [itemct] Item(s) - $[F.cost]<br>
+						[F.desc]<br>"}
+						if(length(F.order_perm))
+							. += "Requires Access (1 or more of)"
+							for(var/acval in F.order_perm)
+								var/accessname = get_access_desc(acval)
+								. += " | [accessname]"
+							. += "<br>"
+						. += "<hr>"
+
+			if(MODE_CART)
+				. += "<h4>Shopping Cart</h4><br>"
+				if(length(src.cart) < 1)
+					. += "Empty - please use Catalogue."
+				else
+					var/entryct = length(src.cart)
+					. += "[entryct] Selections - [cartsize] Items - $[cartcost]<br>"
+					. += "<a href='byond://?src=\ref[src];checkout=1'>Check Out</a> | <a href='byond://?src=\ref[src];clearcart=1'>Clear Cart</a><br><hr>"
+					for(var/P in src.cart)
+						var/datum/mail_order/F = P
+						if(!istype(F, /datum/mail_order))
+							continue
+						var/itemct = length(F.order_items)
+						var/requiresid
+						if(length(F.order_perm))
+							requiresid = " | Requires ID"
+						else
+							requiresid = ""
+						. += {"[F.name]<br>
+						[itemct] Items | $[F.cost][requiresid]<br><hr>"}
+
+	Topic(href, href_list)
+		if(..())
+			return
+
+		if (href_list["viewlist"])
+			src.mode = MODE_LIST
+
+		if (href_list["viewcart"])
+			src.mode = MODE_CART
+
+		if (href_list["checkout"])
+			if(length(src.cart) > 0)
+				var/alert_beep = null
+				if(!src.master.host_program.message_silent)
+					alert_beep = src.master.host_program.message_tone
+
+				if (signal_loss >= 25)
+					src.master.display_alert(alert_beep)
+					var/displayMessage = "[bicon(master)] Unable to place order due to connection failure. Please try again later."
+					src.master.display_message(displayMessage)
+
+				else
+					var/creditCheck = src.authCard(usr)
+					if(creditCheck == "SUCCESS")
+						var/destination = "Send to QM"
+						if(pick_landmark(LANDMARK_MAILORDER_SPAWN)) //pick a destination if mail insertion is supported by map
```

Comment:
I don't see a reason to pick randomly here? Alternatively it might be a good idea to add some helper proc like `landmark_exists` to landmark.dm.
```suggestion
						if(length(landmarks?[LANDMARK_MAILORDER_SPAWN])) //pick a destination if mail insertion is supported by map
```

Reply:
Lengthified.